File: js/openaiResponses/multimodal.ts
import "dotenv/config";
// @ts-ignore
import fs from "fs";
import { wrapOpenAI } from "langsmith/wrappers";
import { OpenAI } from "openai";
import { image } from "../resources/base64BufoApproval";
import { milwaukeeArtMuseumPDFBase64 } from "../resources/Milwaukee Art Museum PDF";

const client = wrapOpenAI(new OpenAI());

async function imageUrlExample() {
  const response = await client.responses.create({
    model: "gpt-4o",
    input: [
      {
        role: "user",
        content: [
          { type: "input_text", text: "what's in this image?" },
          {
            detail: "auto",
            type: "input_image",
            image_url:
              "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
          },
        ],
      },
    ],
  });

  console.log("Image URL response:");
  console.log(response.output_text);
}

async function imageBase64Example() {
  const response = await client.responses.create({
    model: "gpt-4.1-mini",
    input: [
      {
        role: "user",
        content: [
          { type: "input_text", text: "what's in this image?" },
          {
            detail: "auto",
            type: "input_image",
            image_url: `${image}`,
          },
        ],
      },
    ],
  });

  console.log("\nImage base64 response:");
  console.log(response.output_text);
}

async function pdfUploadExample() {
  const file = await client.files.create({
    file: fs.createReadStream("./resources/Milwaukee Art Museum Summary.pdf"),
    purpose: "user_data",
  });

  const response = await client.responses.create({
    model: "gpt-4o",
    input: [
      {
        role: "user",
        content: [
          { type: "input_text", text: "what's in this pdf?" },
          {
            type: "input_file",
            file_id: file.id,
          },
        ],
      },
    ],
  });

  console.log("\nPDF upload response:");
  console.log(response.output_text);
}

async function pdfBase64Example() {
  const response = await client.responses.create({
    model: "gpt-4o",
    input: [
      {
        role: "user",
        content: [
          { type: "input_text", text: "what's in this pdf?" },
          {
            type: "input_file",
            filename: "milwaukeeArtMuseum.pdf",
            file_data: `data:application/pdf;base64,${milwaukeeArtMuseumPDFBase64}`,
          },
        ],
      },
    ],
  });

  console.log("\nPDF base64 response:");
  console.log(response.output_text);
}
async function pdfUrlExample() {
  const response = await client.responses.create({
    model: "gpt-4o",
    input: [
      {
        role: "user",
        content: [
          { type: "input_text", text: "what's in this pdf?" },
          {
            type: "input_file",
            file_url:
              "https://drive.google.com/uc?export=download&id=1m0_e3dOKrbLY3VNjnTehQ3hR-4t5V4FB",
          },
        ],
      },
    ],
  });

  console.log("Image URL response:");
  console.log(response.output_text);
}

async function multimodalExample() {
  await imageUrlExample();
  await imageBase64Example();
  await pdfUploadExample();
  await pdfBase64Example();
  await pdfUrlExample();
}

export async function main() {
  return multimodalExample();
}

File: python/wrap_openai_responses/multimodal.py
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))

import openai
from helpers import fetch_image_base64, fetch_pdf_base64
from langsmith import traceable
from langsmith.wrappers import wrap_openai

client = wrap_openai(openai.Client())


@traceable(name="Responses API Image URL")
def image_url_example():
    """Image input using URL."""
    response = client.responses.create(
        model="gpt-4o",
        input=[
            {
                "role": "user",
                "content": [
                    {
                        "type": "input_text",
                        "text": "Come up with keywords related to the image, and search on the web using the search tool for any news related to the keywords, summarize the findings and cite the sources.",
                    },
                    {
                        "type": "input_image",
                        "image_url": "https://upload.wikimedia.org/wikipedia/commons/thumb/1/15/Cat_August_2010-4.jpg/2880px-Cat_August_2010-4.jpg",
                    },
                ],
            }
        ],
        tools=[{"type": "web_search"}],
    )

    print(f"Image URL response: {response.output_text[:100]}...")
    return response


@traceable(name="Responses API Image Base64")
def image_base64_example():
    """Image input using base64."""
    image_url = "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg"
    image_data = fetch_image_base64(image_url)

    response = client.responses.create(
        model="gpt-4.1-mini",
        input=[
            {
                "role": "user",
                "content": [
                    {"type": "input_text", "text": "what's in this image?"},
                    {
                        "detail": "auto",
                        "type": "input_image",
                        "image_url": f"data:image/jpeg;base64,{image_data}",
                    },
                ],
            }
        ],
    )

    print(f"Image base64 response: {response.output_text[:100]}...")
    return response


@traceable(name="Responses API PDF Base64")
def pdf_base64_example():
    """PDF input using base64."""
    pdf_url = "https://pdfobject.com/pdf/sample.pdf"
    pdf_data = fetch_pdf_base64(pdf_url)

    response = client.responses.create(
        model="gpt-4o",
        input=[
            {
                "role": "user",
                "content": [
                    {"type": "input_text", "text": "what's in this pdf?"},
                    {
                        "type": "input_file",
                        "filename": "sample.pdf",
                        "file_data": f"data:application/pdf;base64,{pdf_data}",
                    },
                ],
            }
        ],
    )

    print(f"PDF response: {response.output_text[:100]}...")
    return response


def main():
    print("Running wrap_openai responses API multimodal examples...")
    image_url_example()
    image_base64_example()
    pdf_base64_example()
    return {"multimodal": "complete"}
